R/domain.R

```diff
@@ -0,0 +1,81 @@
+#' @include utils-pipe.R
+NULL
+
+domain <- function(client, domain_name) {
+  protocol <- client$.__protocol__
+  members_names <- names(c(
+    protocol$list_commands(domain_name),
+    protocol$list_events(domain_name),
+    protocol$list_types(domain_name)
+  ))
+
+  CustomDomain <- R6::R6Class(
+    domain_name,
+    inherit = Domain,
+    public = list(
+      initialize = function(client, domain) {
+        super$initialize(client, domain)
+        protocol <- client$.__protocol__
+        commands <- protocol$list_commands(domain)
+        events <- protocol$list_events(domain)
+        types <- protocol$list_types(domain)
+        purrr::iwalk(commands, function(command, name) {
+          self[[name]] <- private$.build_command(command, name)
+        })
+        purrr::iwalk(events, function(event, name) {
+          self[[name]] <- private$.build_event_listener(event)
+        })
+      }
+    )
+  )
+
+  purrr::walk(members_names, ~ CustomDomain$set("public", .x, NULL))
+  CustomDomain$new(client, domain_name)
+}
+
+Domain <- R6::R6Class(
+  "Domain",
+  public = list(
+    initialize = function(client, domain) {
+      self$.__client__ <- client
+      private$.domain_name <- domain
+    },
+    .__client__ = NULL
+  ),
+  private = list(
+    .domain_name = NULL,
+    .build_command = function(method_to_be_sent, name) {
+      fun <- function() {
+        rlang::fn_fmls_names() %>% # pick the fun arguments
+          utils::head(-1) %>% # remove the callback argument
+          rlang::env_get_list(nms = ., inherit = TRUE) %>% # retrieve values
+          purrr::discard(~ purrr::is_null(.x)) %>% # remove arguments identical to NULL
+          self$.__client__$send(
+            # since the function parameters are not controlled,
+            # there might be some conflicts between CDP parameters and `method_to_be_sent`
+            # Therefore, use get() to retrieve the `method_to_be_sent`
+            get("method_to_be_sent", envir = parent.env(environment())),
```

Comment:
I think you could stick with `rlang` here:

* look directly in the env we know it is there  
```r
rlang::env_get(env = rlang::env_parent(), nm = "method_to_be_sent")
```

* or just look inside every parent env too 
```r
rlang::env_get(nm = "method_to_be_sent", inherit = TRUE)
```

If we want to use `rlang` for env manipulation, I think it is better to use it everywhere

Reply:
Issue opened to deal with that after merged (#44)